### DISCUSSION TITLE
소프트웨어 개발을 하다 보면 코드를 “추상화”하는 경우를 만나게 됩니다

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#67] [프론트 기술 질문 공간]
1. 소프트웨어 개발에서 “추상화”라고 하는 개념에 대해 소개해주세요.
    1. 코드를 추상화하면 어떤 장점이 있나요?
    2. 본인이 잘 추상화했다고 생각하는 코드 조각 하나를 큰 틀에서 소개해주세요. 왜 잘 추상화했다고 생각하셨나요?
    3. 본인이 잘못 추상화했다고 생각하는 코드 조각 하나를 소개해주세요. 왜 나쁘다고 생각하셨나요?

**Comment by javascriptbiggosoo** [comment]
1. 내부 구현에서 여러 절차를 거쳐야하는 코드를 한 덩이의 함수로 만들어 둘 수 있다면 함수의 이름만 가지고도 해당 코드가 어떤 역할을 하는지 파악하기 쉬울 것입니다.
2. +3. 코드 안에서 추상화 수준은 비슷해야합니다. 높은 추상화와 낮은 추상화가 섞여있다면 한눈에 전체 코드를 파악하기 힘듭니다.
```jsx
// 좋은 예시
function UserProfile({ user }) {
  return (
    <div>
      <UserHeader user={user} />
      <UserBody user={user} />
    </div>
  );
}

function UserHeader({ user }) {
  return <h2>{user.name}</h2>;
}

function UserBody({ user }) {
  return <p>{user.description}</p>;
}
```
```jsx
// 나쁜 예시
function UserProfile({ user }) {
  return (
    <div>
      <h2>{user.name}</h2>
      ...
      <UserBody user={user} />
    </div>
  );
}

function UserBody({ user }) {
  return <p>{user.description}</p>;
}
```
좋은 예시의 경우 레이아웃의 각 영역을 따로 리액트 컴포넌트를 만들어 비슷한 수준으로 추상화했습니다.
나쁜 예시에서는 비슷한 수준의 레벨에서 한쪽은 컴포넌트화를 했고 한쪽은 하지 않음으로서 추상화의 일관성이 떨어뜨려 전체적인 구조를 한눈에 파악하기가 힘듭니다.


이 외에도 하나의 함수가 두가지 이상의 책임을 가지지 않도록 하는 단일책임원칙 또한 추상화에서 신경써야할 중요한 요소입니다.

**Comment by michoball** [comment]
1. 추상화는 객체지향 프로그래밍에서 주로 사용되는 개념으로 객체의 공통적인 속성과 기능을 추출하여 정의하는것을 의미합니다. 
2.
```typescript
interface BackdropProps {
  onToggleModal?: () => void
}

interface ModalWrapperProps {
  children: React.ReactNode
}

interface ModalProps extends ModalWrapperProps {
  toggleModal?: () => void
}

export const Backdrop: React.FC<BackdropProps> = ({ onToggleModal }) => {
  return <BackdropStyle onClick={onToggleModal}></BackdropStyle>
}

const ModalWrapper: React.FC<ModalWrapperProps> = ({ children }) => {
  return <ModalWrapperStyle>{children}</ModalWrapperStyle>
}

const Modal: React.FC<ModalProps> = ({ toggleModal, children }) => {
  return (
    <>
      {createPortal(
        <Backdrop onToggleModal={toggleModal} />,
        document.getElementById('backdorp') as HTMLDivElement
      )}
      {createPortal(
        <ModalWrapper>{children}</ModalWrapper>,
        document.getElementById('overlay') as HTMLDivElement
      )}
    </>
  )
}

export default Modal
```

3. 잘 모르겠습니다.

**Comment by BaekYuri** [comment]
소프트웨어 개발에서 “추상화”라고 하는 개념에 대해 소개해주세요.
 -> 어떠한 객체의 행동(메소드)을 메소드의 이름만으로도 쉽게 유추할 수 있도록 단순화하고 일반화 시키는 것입니다.
1. 코드를 추상화하면 어떤 장점이 있나요?
 -> 추상화를 통해 이해하기 어려운 객체의 특성에 쉽게 다가갈 수 있다고 생각합니다.
2. 본인이 잘 추상화했다고 생각하는 코드 조각 하나를 큰 틀에서 소개해주세요. 왜 잘 추상화했다고 생각하셨나요?
 -> 메소드 이름만으로도 순열에 대한 기능을 한다는 것을 알 수 있습니다.
```java
static void permutation(int idx, int state, int[] now) {
		if(idx==N-1) {
			min = Math.min(min, state);
			max = Math.max(max, state);
			return;
		}
		for(int i=0;i<4;i++) {
			if(now[i]==0) continue;
			now[i]--;
			switch(i) {
			case 0:	
				permutation(idx+1, state+numbers[idx+1],now);
				break;
			case 1:	
				permutation(idx+1, state-numbers[idx+1],now);	
				break;
			case 2:	
				permutation(idx+1, state*numbers[idx+1],now);		
				break;
			case 3:			
				permutation(idx+1, state/numbers[idx+1],now);
				break;
				
			}
			now[i]++;
		}
	}
```
3. 본인이 잘못 추상화했다고 생각하는 코드 조각 하나를 소개해주세요. 왜 나쁘다고 생각하셨나요?
 -> 무엇을 만드는 함수인 것은 알겠지만, 정확히 무엇을 만드는지 메소드 이름만으로 추측이 가지 않습니다.
```java
static void make() {
		parents = new int[length*length];
		rank = new int[length*length];
		for(int i=0;i<parents.length;i++) {
			parents[i] = -1;
		}
	}
```

**Comment by WooGie911** [comment]
소프트웨어 개발에서 “추상화”라고 하는 개념에 대해 소개해주세요.

=>

추상화 : 하나의 기능을 하나의 함수로 만드는 작업 
 -> 하나의 함수는 한가지의 관심사(추상화)를 담는다.

추상화 수준에 따라 추상화 수준이 높은 추상화와 추상화 수준이 낮은 추상화로 나뉜다.
 추상화 수준이 높은 추상화 : 함수로만 구현되어있는 코드.
 추상화 수준이 낮은 추상화 : 연산기호나 형변환같은 구현사항이 담겨있는 코드.

내려가기 규칙 : 추상화 수준이 높은 추상화를 위에, 추상화 수준이 낮은 추상화를 아래에 작성하는 것 


1. 코드를 추상화하면 어떤 장점이 있나요?
 -> 추상화를 잘 지킨 코드 : 클린한 코드 = 유지 보수에 유리.
